### PR TITLE
APIs and CRDs for backup object

### LINKED ISSUE #31: Implement API and generate CRD's for backup
**Issue/Feature Description:**

**Why this issue to fixed / feature is needed(give scenarios or use cases):**
1. Implement API spec
2. Generate CRD's 
3. generate clientset/informers etc.

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by PravinRanjan10**
**What type of PR is this?**
This PR is a new feature for Kahu backup. It contains the Backup API and CRD and generated client

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #31 

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:

**Special notes for your reviewer**:
How to generate API-Client code and CRDs:
1. code-generator/generate-groups.sh all github.com/soda-cdm/kahu/controllers/generated/client github.com/soda-cdm/kahu/apis kahu:v1beta1 --go-header-file ../kahu1/vendor/k8s.io/code-generator/hack/boilerplate.go.txt

2. controller-gen paths=github.com/soda-cdm/kahu/apis/kahu/v1beta1 crd:crdVersions=v1 output:crd:artifacts:config=config

### CODE REVIEW

**Comment by PravinRanjan10** [comment]
@sushanthakumar @joseph-v @AmitRoushan plz review

**Comment by AmitRoushan** [comment]
@PravinRanjan10 Please create an issue from the PR

File: apis/kahu/v1beta1/backup.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	// +optional
+	ReclaimPolicyRetain string `json:"reclaimPolicyRetain,omitempty"`
+}
+
+// BackupCondition indicates the current state of a resource that is backing up
+type BackupCondition struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (AmitRoushan) [inline apis/kahu/v1beta1/backup.go:138]: IMO, ```metav1.TypeMeta``` and ```namespaceName``` are sufficient for ```BackupCondition```.

File: apis/kahu/v1beta1/backup.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	Status string `json:"status,omitempty"`
+}
+
+// BackupPhase is a state of backup
+// +kubebuilder:validation:Enum=New;FailedValidation;InProgress;Completed;PartiallyFailed;Failed;Deleting
+type BackupPhase string
+
+const (
+	// BackupPhaseNew indicates that current backup object is New
+	BackupPhaseInit BackupPhase = "New"
```
Comment (AmitRoushan) [inline apis/kahu/v1beta1/backup.go:155]: Comment and Constant name donot match

File: config/kahu.io_backupconditions.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+---
```
Comment (AmitRoushan) [inline config/kahu.io_backupconditions.yaml:1]: IMHO, ```Backupconditions``` CRD schema canot not be CRD as it is part of ```Backup``` resource status

File: apis/kahu/v1beta1/backup_location.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BackupLocationSpec defines the desired state of BackupLocation
```
Comment (AmitRoushan) [inline apis/kahu/v1beta1/backup_location.go:21]: Client code generation tags are missing

File: apis/kahu/v1beta1/backup_location.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// BackupLocation is the Schema for the backuplocations API
+type BackupLocation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec BackupLocationSpec `json:"spec,omitempty"`
+}
```
Comment (AmitRoushan) [inline apis/kahu/v1beta1/backup_location.go:39]: ```Status``` might be required for the ```BackupLocation``` Resources

**Review by AmitRoushan — CHANGES_REQUESTED** [review]

**Comment by AmitRoushan** [comment]
@PravinRanjan10 Please add doc/script to generate CRDs yaml, clientset and configuration

File: apis/kahu/v1beta1/backup.go (hunk 2 above)
Comment (PravinRanjan10) [reply apis/kahu/v1beta1/backup.go:152]: done

File: apis/kahu/v1beta1/backup_location.go (hunk 5 above)
Comment (PravinRanjan10) [reply apis/kahu/v1beta1/backup_location.go:44]: For adding this, we may need some discussion.So we can add it later.

File: apis/kahu/v1beta1/backup.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	// +optional
+	MetadataLocation *BackupLocation `json:"metadataLocation,omitempty"`
+
+	// ReclaimPolicy tells about reclamation of the backup. It can be either delete or retain
+	// +optional
+	ReclaimPolicy ReclaimPolicyType `json:"reclaimPolicy,omitempty"`
+
+	// PreExecHook is a hook which should be executed before backup starts
+	// +optional
+	PreExecHook ResourceHookSpec `json:"preExecHook,omitempty"`
```
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply apis/kahu/v1beta1/backup.go:49]: I think it's ok, as we are sure these parameters will be added for backup.

File: apis/kahu/v1beta1/backup_location.go (hunk 4 above)
Comment (PravinRanjan10) [reply apis/kahu/v1beta1/backup_location.go:21]: done

**Review by sushanthakumar — APPROVED** [review]

**Comment by AmitRoushan** [comment]
@PravinRanjan10 IMO, informer, lister and client set generated files require SODA Authors in header instead of Kubernetes.

**Comment by PravinRanjan10** [comment]
> @PravinRanjan10 IMO, informer, lister and client set generated files require SODA Authors in header instead of Kubernetes.

Done

**Review by sushanthakumar — APPROVED** [review]

**Review by AmitRoushan — APPROVED** [review]
LGTM

**Pull request merged**